tests/core/utilities/test_abi.py

```diff
@@ -29,9 +34,45 @@ def test_abi_data_tree(types, data, expected):
     assert abi_data_tree(types, data) == expected
 
 
+@pytest.mark.parametrize(
+    'func, data_tree, expected',
+    [
+        (
+            addresses_checksummed,
+            [
+                ABITypedData(
+                    [
+                        'address',
+                        b'\xf2\xe2F\xbbv\xdf\x87l\xef\x8b8\xae\x84\x13\x0fOU\xde9[',
+                    ]
+                ),
+                ABITypedData([None, 'latest'])
+            ],
+            [
+                ABITypedData(
+                    [
+                        'address',
+                        '0xF2E246BB76DF876Cef8b38ae84130F4F55De395b',
+                    ]
+                ),
+                ABITypedData([None, 'latest'])
+            ]
+        )
+    ],
+)
+def test_data_tree_map(func, data_tree, expected):
+    assert data_tree_map(func, data_tree) == expected
+
+
 @pytest.mark.parametrize(
     'types, data, funcs, expected',
     [
+        (  # like web3._utils.rpc_abi.RPC_ABIS['eth_getCode']
+            ['address', None],
```

Comment:
This has come up a few times, so let's all be on the lookout for it: recursive structures must be handled. For one example, what happens when we try to run a map function over ABI data that has an array inside a tuple?

Something like:
```py
        (
            ['(string,address[])'],
            [(b'a string', [b'\xf2\xe2F\xbbv\xdf\x87l\xef\x8b8\xae\x84\x13\x0fOU\xde9['],
            [addresses_checksummed, abi_string_to_text],
            ['a string', ['0xF2E246BB76DF876Cef8b38ae84130F4F55De395b']],
        ),